File: frontend/src/api/endpoints/institucional.ts
import { api } from '@/api/client';
import type {
  Campanha,
  CampanhaStatus,
  ID,
  PedidoColeta,
  PedidoColetaStatus,
} from '@/types/api';

export const pedidosColetaApi = {
  list: () => api.get<PedidoColeta[]>('/pedidos-coleta/').then((r) => r.data),
  criar: (payload: {
    bairro: string;
    cidade: string;
    uf: string;
    ibge_municipio?: string;
    tipo_residuo: string;
    foto_path?: string;
    quantidade_estimada?: number | string;
    descricao?: string;
    localizacao: { lat: number; lng: number };
  }) => api.post<PedidoColeta>('/pedidos-coleta/', payload).then((r) => r.data),
  atualizarStatus: (id: ID, status: PedidoColetaStatus, nota?: string) =>
    api
      .patch<PedidoColeta>(`/pedidos-coleta/${id}/status`, { status, nota })
      .then((r) => r.data),
  contestar: (id: ID) =>
    api.post<PedidoColeta>(`/pedidos-coleta/${id}/contestar`).then((r) => r.data),
};

export const campanhasApi = {
  list: (ibge?: string) =>
    api.get<Campanha[]>('/campanhas/', { params: { ibge } }).then((r) => r.data),
  criar: (payload: {
    titulo: string;
    descricao: string;
    data_evento?: string;
    tipo_residuo?: string;
    beneficio?: string;
    cidade?: string;
    uf?: string;
    ibge_municipio?: string;
  }) => api.post<Campanha>('/campanhas/', payload).then((r) => r.data),
  atualizar: (id: ID, payload: Partial<Campanha> & { status?: CampanhaStatus }) =>
    api.patch<Campanha>(`/campanhas/${id}`, payload).then((r) => r.data),
};

export const mapaInstitucionalApi = {
  consultar: (ibge: string) =>
    api
      .get<{
        territorio: string;
        celulas: Array<{
          bairro: string;
          ibge_municipio: string | null;
          anuncios_ativos: number;
          pedidos_abertos: number;
          campanhas_ativas: number;
        }>;
      }>('/mapa-institucional/', { params: { ibge } })
      .then((r) => r.data),
};

File: backend/app/schemas/institucional.py
"""Schemas — PedidoColetaPublica, CampanhaPublica."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, Field, StringConstraints

from app.models.enums import CampanhaStatus, PedidoColetaStatus
from app.schemas.common import GeoPoint, ORMModel, TimestampedORM


# === PedidoColetaPublica ===

class PedidoColetaCreate(BaseModel):
    bairro: Annotated[str, StringConstraints(max_length=120)]
    cidade: Annotated[str, StringConstraints(max_length=100)]
    uf: Annotated[str, StringConstraints(pattern=r"^[A-Z]{2}$")]
    ibge_municipio: Annotated[str, StringConstraints(pattern=r"^\d{7}$")] | None = None
    tipo_residuo: Annotated[str, StringConstraints(max_length=120)]
    foto_path: str | None = None
    quantidade_estimada: Decimal | None = None
    descricao: str | None = None
    localizacao: GeoPoint


class PedidoColetaRead(TimestampedORM):
    conta_solicitante_id: UUID
    prefeitura_conta_id: UUID | None
    bairro: str
    cidade: str
    uf: str
    ibge_municipio: str | None
    tipo_residuo: str
    foto_path: str | None
    quantidade_estimada: Decimal | None
    descricao: str | None
    lat: float
    lng: float
    status: PedidoColetaStatus


class PedidoColetaStatusUpdate(BaseModel):
    status: PedidoColetaStatus
    nota: str | None = None


# === CampanhaPublica ===

class CampanhaCreate(BaseModel):
    titulo: Annotated[str, StringConstraints(min_length=3, max_length=200)]
    descricao: Annotated[str, StringConstraints(min_length=10)]
    data_evento: datetime | None = None
    tipo_residuo: str | None = None
    beneficio: str | None = None
    cidade: str | None = None
    uf: str | None = None
    ibge_municipio: str | None = None


class CampanhaUpdate(BaseModel):
    titulo: str | None = None
    descricao: str | None = None
    data_evento: datetime | None = None
    tipo_residuo: str | None = None
    beneficio: str | None = None
    status: CampanhaStatus | None = None


class CampanhaRead(TimestampedORM):
    conta_organizadora_id: UUID
    titulo: str
    descricao: str
    data_evento: datetime | None
    tipo_residuo: str | None
    beneficio: str | None
    cidade: str | None
    uf: str | None
    ibge_municipio: str | None
    status: CampanhaStatus


# === MapaInstitucional ===

class MapaInstitucionalCelula(ORMModel):
    bairro: str
    cidade: str
    uf: str
    ibge_municipio: str | None
    anuncios_ativos: int = Field(ge=0)
    pedidos_abertos: int = Field(ge=0)
    campanhas_ativas: int = Field(ge=0)


class MapaInstitucionalResposta(ORMModel):
    territorio: str  # "municipio:3550308" | "estado:SP"
    celulas: list[MapaInstitucionalCelula]
